File: desktop-app-vue/src/main/ai-engine/content-recommender.js
const { logger, createLogger } = require('../utils/logger.js');

/**
 * ContentRecommender - 基于内容的推荐算法
 * P2智能层Phase 4 - 推荐系统
 *
 * 功能:
 * - 工具特征提取
 * - 工具相似度计算
 * - 基于内容的推荐
 * - 工具链推荐
 *
 * Version: v0.24.0
 * Date: 2026-01-02
 */

class ContentRecommender {
  constructor(config = {}) {
    this.config = {
      minSimilarity: 0.2,      // 最小相似度阈值
      topKSimilar: 5,          // 考虑Top-K相似工具
      enableToolChain: true,   // 启用工具链推荐
      ...config
    };

    this.db = null;
    this.toolFeatures = new Map();     // 工具特征映射
    this.toolSimilarity = new Map();   // 工具相似度矩阵
    this.toolChains = new Map();       // 工具链统计

    this.stats = {
      totalRecommendations: 0,
      toolChainRecommendations: 0,
      avgSimilarity: 0
    };
  }

  /**
   * 设置数据库连接
   */
  setDatabase(db) {
    this.db = db;
  }

  /**
   * 构建工具特征
   */
  async buildToolFeatures() {
    logger.info('[ContentRecommender] 构建工具特征...');

    // 默认工具特征库
    const defaultTools = [
      {
        name: 'codeGeneration',
        category: 'development',
        tags: ['code', 'generate', 'function', 'class'],
        description: '代码生成工具，自动生成函数、类等代码结构'
      },
      {
        name: 'fileWrite',
        category: 'development',
        tags: ['file', 'write', 'save', 'create'],
        description: '文件写入工具，创建和保存文件内容'
      },
      {
        name: 'fileRead',
        category: 'development',
        tags: ['file', 'read', 'load', 'open'],
        description: '文件读取工具，加载和读取文件内容'
      },
      {
        name: 'formatCode',
        category: 'development',
        tags: ['format', 'code', 'style', 'beautify'],
        description: '代码格式化工具，美化代码风格'
      },
      {
        name: 'debugging',
        category: 'development',
        tags: ['debug', 'fix', 'error', 'troubleshoot'],
        description: '调试工具，帮助排查和修复代码错误'
      },
      {
        name: 'testing',
        category: 'development',
        tags: ['test', 'unittest', 'qa', 'verify'],
        description: '测试工具，编写和运行单元测试'
      },
      {
        name: 'dataAnalysis',
        category: 'data',
        tags: ['data', 'analysis', 'statistics', 'explore'],
        description: '数据分析工具，统计分析和探索数据'
      },
      {
        name: 'chartGeneration',
        category: 'data',
        tags: ['chart', 'visualization', 'graph', 'plot'],
        description: '图表生成工具，可视化数据生成图表'
      },
      {
        name: 'documentation',
        category: 'writing',
        tags: ['doc', 'document', 'write', 'readme'],
        description: '文档编写工具，生成项目文档和说明'
      },
      {
        name: 'markdown',
        category: 'writing',
        tags: ['markdown', 'format', 'text', 'note'],
        description: 'Markdown编辑工具，格式化文本内容'
      }
    ];

    this.toolFeatures.clear();
    for (const tool of defaultTools) {
      this.toolFeatures.set(tool.name, {
        category: tool.category,
        tags: new Set(tool.tags),
        description: tool.description,
        vector: this.createFeatureVector(tool)
      });
    }

    logger.info(`[ContentRecommender] 工具特征构建完成: ${this.toolFeatures.size}个工具`);
    return this.toolFeatures;
  }

  /**
   * 创建工具特征向量
   */
  createFeatureVector(tool) {
    const vector = [];

    // 1. 类别编码
    const categories = ['development', 'data', 'design', 'writing', 'testing', 'deployment'];
    const categoryIndex = categories.indexOf(tool.category);
    for (let i = 0; i < categories.length; i++) {
      vector.push(i === categoryIndex ? 1 : 0);
    }

    // 2. 标签特征 (TF-IDF简化版本)
    const allTags = ['code', 'file', 'data', 'test', 'doc', 'format', 'debug', 'chart'];
    for (const tag of allTags) {
      vector.push(tool.tags.includes(tag) ? 1 : 0);
    }

    return vector;
  }

  /**
   * 计算工具相似度 (余弦相似度)
   * @param {string} tool1 - 工具1
   * @param {string} tool2 - 工具2
   * @returns {number} 相似度 [0, 1]
   */
  calculateToolSimilarity(tool1, tool2) {
    const cacheKey = `${tool1}:${tool2}`;
    if (this.toolSimilarity.has(cacheKey)) {
      return this.toolSimilarity.get(cacheKey);
    }

    const features1 = this.toolFeatures.get(tool1);
    const features2 = this.toolFeatures.get(tool2);

    if (!features1 || !features2) {
      return 0;
    }

    // 1. 类别相似度
    const categorySimilarity = features1.category === features2.category ? 1 : 0;

    // 2. 标签相似度 (Jaccard)
    const tags1 = features1.tags;
    const tags2 = features2.tags;
    const intersection = new Set([...tags1].filter(t => tags2.has(t)));
    const union = new Set([...tags1, ...tags2]);
    const tagSimilarity = union.size > 0 ? intersection.size / union.size : 0;

    // 3. 特征向量余弦相似度
    const vector1 = features1.vector;
    const vector2 = features2.vector;

    let dotProduct = 0;
    let norm1 = 0;
    let norm2 = 0;

    for (let i = 0; i < vector1.length; i++) {
      dotProduct += vector1[i] * vector2[i];
      norm1 += vector1[i] * vector1[i];
      norm2 += vector2[i] * vector2[i];
    }

    const vectorSimilarity = (norm1 === 0 || norm2 === 0)
      ? 0
      : dotProduct / (Math.sqrt(norm1) * Math.sqrt(norm2));

    // 加权平均
    const similarity = categorySimilarity * 0.4 + tagSimilarity * 0.3 + vectorSimilarity * 0.3;

    // 缓存结果
    this.toolSimilarity.set(cacheKey, similarity);
    this.toolSimilarity.set(`${tool2}:${tool1}`, similarity); // 对称性

    return similarity;
  }

  /**
   * 查找相似工具
   * @param {string} toolName - 工具名称
   * @returns {Array} 相似工具列表 [{tool, similarity}]
   */
  findSimilarTools(toolName) {
    if (this.toolFeatures.size === 0) {
      this.buildToolFeatures();
    }

    const similarTools = [];

    for (const [otherTool, _] of this.toolFeatures.entries()) {
      if (otherTool === toolName) {continue;}

      const similarity = this.calculateToolSimilarity(toolName, otherTool);

      if (similarity >= this.config.minSimilarity) {
        similarTools.push({ tool: otherTool, similarity });
      }
    }

    similarTools.sort((a, b) => b.similarity - a.similarity);
    return similarTools.slice(0, this.config.topKSimilar);
  }

  /**
   * 构建工具链统计
   */
  async buildToolChains() {
    if (!this.db) {return;}

    logger.info('[ContentRecommender] 构建工具链统计...');

    try {
      // 查询工具序列 (previous_tool → tool_name → next_tool)
      const chains = this.db.prepare(`
        SELECT
          previous_tool,
          tool_name,
          COUNT(*) as count
        FROM tool_usage_events
        WHERE previous_tool IS NOT NULL
        GROUP BY previous_tool, tool_name
        ORDER BY count DESC
      `).all();

      this.toolChains.clear();

      for (const chain of chains) {
        if (!this.toolChains.has(chain.previous_tool)) {
          this.toolChains.set(chain.previous_tool, []);
        }

        this.toolChains.get(chain.previous_tool).push({
          nextTool: chain.tool_name,
          count: chain.count,
          probability: 0 // 后续计算
        });
      }

      // 计算条件概率
      for (const [prevTool, nextTools] of this.toolChains.entries()) {
        const totalCount = nextTools.reduce((sum, t) => sum + t.count, 0);
        for (const tool of nextTools) {
          tool.probability = tool.count / totalCount;
        }
      }

      logger.info(`[ContentRecommender] 工具链构建完成: ${this.toolChains.size}个起始工具`);
    } catch (error) {
      logger.error('[ContentRecommender] 构建工具链失败:', error);
    }
  }

  /**
   * 基于内容推荐工具
   * @param {string} userId - 用户ID
   * @param {number} topK - 推荐数量
   * @returns {Array} 推荐列表
   */
  async recommendTools(userId, topK = 5) {
    if (!this.db) {
      return [];
    }

    if (this.toolFeatures.size === 0) {
      await this.buildToolFeatures();
    }

    try {
      // 1. 获取用户最近使用的工具
      const recentTools = this.db.prepare(`
        SELECT DISTINCT tool_name, MAX(timestamp) as last_used
        FROM tool_usage_events
        WHERE user_id = ?
          AND timestamp >= datetime('now', '-7 days')
        GROUP BY tool_name
        ORDER BY last_used DESC
        LIMIT 5
      `).all(userId);

      if (recentTools.length === 0) {
        logger.info('[ContentRecommender] 用户无最近使用记录');
        return [];
      }

      // 2. 基于相似度推荐
      const toolScores = new Map();

      for (const { tool_name } of recentTools) {
        const similarTools = this.findSimilarTools(tool_name);

        for (const { tool, similarity } of similarTools) {
          if (!toolScores.has(tool)) {
            toolScores.set(tool, {
              totalScore: 0,
              count: 0,
              basedOnTools: []
            });
          }

          const data = toolScores.get(tool);
          data.totalScore += similarity;
          data.count++;
          data.basedOnTools.push({ tool: tool_name, similarity });
        }
      }

      // 3. 计算最终评分
      const recommendations = [];
      for (const [tool, data] of toolScores.entries()) {
        const avgScore = data.totalScore / data.count;

        recommendations.push({
          tool,
          score: avgScore,
          confidence: Math.min(data.count / recentTools.length, 1.0),
          reason: this.generateReason(data.basedOnTools),
          algorithm: 'content_based'
        });
      }

      recommendations.sort((a, b) => b.score - a.score);

      this.stats.totalRecommendations++;
      if (recommendations.length > 0) {
        this.stats.avgSimilarity =
          (this.stats.avgSimilarity * (this.stats.totalRecommendations - 1) +
            recommendations[0].score) / this.stats.totalRecommendations;
      }

      return recommendations.slice(0, topK);
    } catch (error) {
      logger.error('[ContentRecommender] 推荐失败:', error);
      return [];
    }
  }

  /**
   * 工具链推荐
   * @param {string} currentTool - 当前工具
   * @param {number} topK - 推荐数量
   * @returns {Array} 推荐列表
   */
  async recommendNextTools(currentTool, topK = 3) {
    if (this.toolChains.size === 0) {
      await this.buildToolChains();
    }

    const nextTools = this.toolChains.get(currentTool);

    if (!nextTools || nextTools.length === 0) {
      return [];
    }

    const recommendations = nextTools.slice(0, topK).map(t => ({
      tool: t.nextTool,
      score: t.probability,
      confidence: t.probability,
      reason: `${(t.probability * 100).toFixed(0)}%用户在${currentTool}后使用`,
      algorithm: 'tool_chain'
    }));

    this.stats.toolChainRecommendations++;

    return recommendations;
  }

  /**
   * 生成推荐理由
   */
  generateReason(basedOnTools) {
    if (basedOnTools.length === 0) {
      return '基于内容推荐';
    }

    const topTool = basedOnTools[0];
    const similarity = (topTool.similarity * 100).toFixed(0);

    return `与${topTool.tool}相似 (${similarity}%)`;
  }

  /**
   * 获取统计信息
   */
  getStats() {
    return {
      ...this.stats,
      toolCount: this.toolFeatures.size,
      toolChainCount: this.toolChains.size,
      avgSimilarity: (this.stats.avgSimilarity * 100).toFixed(1) + '%'
    };
  }
}

module.exports = ContentRecommender;

File: ios-app/ChainlessChain/Features/AI/Advanced/ContentRecommender.swift
import Foundation
import CoreCommon

// MARK: - ContentRecommender
/// Content-based recommendation algorithm for tools
/// Provides tool similarity calculation and recommendation
///
/// Features:
/// - Tool feature extraction
/// - Tool similarity calculation (Jaccard + Cosine)
/// - Content-based recommendation
/// - Tool chain recommendation
///
/// Version: v1.7.0
/// Date: 2026-02-10

// MARK: - Tool Features

/// Tool feature representation
struct ToolFeatures {
    let name: String
    let category: ToolCategory
    let tags: Set<String>
    let description: String
    let vector: [Float]
}

/// Tool category
enum ToolCategory: String, Codable, CaseIterable {
    case development
    case data
    case design
    case writing
    case testing
    case deployment
    case unknown

    var index: Int {
        return ToolCategory.allCases.firstIndex(of: self) ?? 0
    }
}

/// Tool similarity result
struct ToolSimilarity {
    let tool: String
    let similarity: Float
}

/// Tool recommendation result
struct ToolRecommendation: Identifiable {
    let id = UUID()
    let tool: String
    let score: Float
    let confidence: Float
    let reason: String
    let algorithm: RecommendationAlgorithm
}

/// Recommendation algorithm type
enum RecommendationAlgorithm: String {
    case contentBased = "content_based"
    case toolChain = "tool_chain"
    case hybrid = "hybrid"
}

/// Tool chain entry
struct ToolChainEntry {
    let nextTool: String
    var count: Int
    var probability: Float
}

// MARK: - Content Recommender

/// Content-based recommendation system for tools
@MainActor
class ContentRecommender: ObservableObject {
    // MARK: - Configuration

    struct Config {
        var minSimilarity: Float = 0.2        // Minimum similarity threshold
        var topKSimilar: Int = 5              // Consider Top-K similar tools
        var enableToolChain: Bool = true       // Enable tool chain recommendation
    }

    // MARK: - Properties

    private var config: Config
    private var toolFeatures: [String: ToolFeatures] = [:]
    private var toolSimilarity: [String: Float] = [:]          // Cache: "tool1:tool2" -> similarity
    private var toolChains: [String: [ToolChainEntry]] = [:]    // Previous tool -> next tools
    private let logger = Logger.shared

    @Published private(set) var stats = RecommenderStats()

    /// Statistics
    struct RecommenderStats {
        var totalRecommendations: Int = 0
        var toolChainRecommendations: Int = 0
        var avgSimilarity: Float = 0
    }

    // Standard tag set for vector building
    private let standardTags = ["code", "file", "data", "test", "doc", "format", "debug", "chart"]

    // MARK: - Singleton

    static let shared = ContentRecommender()

    // MARK: - Initialization

    init(config: Config = Config()) {
        self.config = config
        buildToolFeatures()
    }

    // MARK: - Public Methods

    /// Build tool features
    @discardableResult
    func buildToolFeatures() -> [String: ToolFeatures] {
        logger.info("[ContentRecommender] Building tool features...")

        // Default tool feature library
        let defaultTools: [(name: String, category: ToolCategory, tags: [String], description: String)] = [
            ("codeGeneration", .development, ["code", "generate", "function", "class"],
             "Code generation tool, automatically generates functions, classes, etc."),
            ("fileWrite", .development, ["file", "write", "save", "create"],
             "File writing tool, creates and saves file content"),
            ("fileRead", .development, ["file", "read", "load", "open"],
             "File reading tool, loads and reads file content"),
            ("formatCode", .development, ["format", "code", "style", "beautify"],
             "Code formatting tool, beautifies code style"),
            ("debugging", .development, ["debug", "fix", "error", "troubleshoot"],
             "Debugging tool, helps troubleshoot and fix code errors"),
            ("testing", .testing, ["test", "unittest", "qa", "verify"],
             "Testing tool, writes and runs unit tests"),
            ("dataAnalysis", .data, ["data", "analysis", "statistics", "explore"],
             "Data analysis tool, statistical analysis and data exploration"),
            ("chartGeneration", .data, ["chart", "visualization", "graph", "plot"],
             "Chart generation tool, visualizes data in charts"),
            ("documentation", .writing, ["doc", "document", "write", "readme"],
             "Documentation tool, generates project docs and descriptions"),
            ("markdown", .writing, ["markdown", "format", "text", "note"],
             "Markdown editing tool, formats text content"),
            ("deployment", .deployment, ["deploy", "release", "publish", "build"],
             "Deployment tool, publishes and releases applications"),
            ("containerization", .deployment, ["docker", "container", "image", "kubernetes"],
             "Containerization tool, manages Docker containers"),
            ("uiDesign", .design, ["ui", "design", "layout", "component"],
             "UI design tool, creates user interface components"),
            ("imageProcessing", .design, ["image", "photo", "edit", "filter"],
             "Image processing tool, edits and filters images")
        ]

        toolFeatures.removeAll()

        for tool in defaultTools {
            let features = ToolFeatures(
                name: tool.name,
                category: tool.category,
                tags: Set(tool.tags),
                description: tool.description,
                vector: createFeatureVector(category: tool.category, tags: tool.tags)
            )
            toolFeatures[tool.name] = features
        }

        logger.info("[ContentRecommender] Tool features built: \(toolFeatures.count) tools")
        return toolFeatures
    }

    /// Add custom tool features
    func addToolFeatures(name: String, category: ToolCategory, tags: [String], description: String) {
        let features = ToolFeatures(
            name: name,
            category: category,
            tags: Set(tags),
            description: description,
            vector: createFeatureVector(category: category, tags: tags)
        )
        toolFeatures[name] = features

        // Clear similarity cache for this tool
        let keysToRemove = toolSimilarity.keys.filter { $0.contains(name) }
        for key in keysToRemove {
            toolSimilarity.removeValue(forKey: key)
        }
    }

    /// Calculate tool similarity (cosine + Jaccard)
    /// - Parameters:
    ///   - tool1: First tool name
    ///   - tool2: Second tool name
    /// - Returns: Similarity score [0, 1]
    func calculateToolSimilarity(_ tool1: String, _ tool2: String) -> Float {
        // Check cache
        let cacheKey = "\(tool1):\(tool2)"
        if let cached = toolSimilarity[cacheKey] {
            return cached
        }

        guard let features1 = toolFeatures[tool1],
              let features2 = toolFeatures[tool2] else {
            return 0
        }

        // 1. Category similarity
        let categorySimilarity: Float = features1.category == features2.category ? 1.0 : 0.0

        // 2. Tag similarity (Jaccard)
        let intersection = features1.tags.intersection(features2.tags)
        let union = features1.tags.union(features2.tags)
        let tagSimilarity: Float = union.isEmpty ? 0 : Float(intersection.count) / Float(union.count)

        // 3. Feature vector cosine similarity
        let vectorSimilarity = cosineSimilarity(features1.vector, features2.vector)

        // Weighted average
        let similarity = categorySimilarity * 0.4 + tagSimilarity * 0.3 + vectorSimilarity * 0.3

        // Cache result (symmetric)
        toolSimilarity[cacheKey] = similarity
        toolSimilarity["\(tool2):\(tool1)"] = similarity

        return similarity
    }

    /// Find similar tools
    /// - Parameter toolName: Tool name
    /// - Returns: List of similar tools sorted by similarity
    func findSimilarTools(_ toolName: String) -> [ToolSimilarity] {
        if toolFeatures.isEmpty {
            buildToolFeatures()
        }

        var similarTools: [ToolSimilarity] = []

        for otherTool in toolFeatures.keys where otherTool != toolName {
            let similarity = calculateToolSimilarity(toolName, otherTool)

            if similarity >= config.minSimilarity {
                similarTools.append(ToolSimilarity(tool: otherTool, similarity: similarity))
            }
        }

        similarTools.sort { $0.similarity > $1.similarity }
        return Array(similarTools.prefix(config.topKSimilar))
    }

    /// Build tool chain statistics from usage history
    /// - Parameter usageHistory: Array of (previousTool, currentTool, count)
    func buildToolChains(from usageHistory: [(previousTool: String, currentTool: String, count: Int)]) {
        logger.info("[ContentRecommender] Building tool chain statistics...")

        toolChains.removeAll()

        for entry in usageHistory {
            if toolChains[entry.previousTool] == nil {
                toolChains[entry.previousTool] = []
            }

            toolChains[entry.previousTool]?.append(ToolChainEntry(
                nextTool: entry.currentTool,
                count: entry.count,
                probability: 0  // Calculate later
            ))
        }

        // Calculate conditional probabilities
        for (prevTool, var nextTools) in toolChains {
            let totalCount = nextTools.reduce(0) { $0 + $1.count }
            for i in 0..<nextTools.count {
                nextTools[i].probability = Float(nextTools[i].count) / Float(totalCount)
            }
            toolChains[prevTool] = nextTools.sorted { $0.probability > $1.probability }
        }

        logger.info("[ContentRecommender] Tool chains built: \(toolChains.count) starting tools")
    }

    /// Content-based tool recommendation
    /// - Parameters:
    ///   - recentTools: User's recently used tools
    ///   - topK: Number of recommendations
    /// - Returns: List of recommendations
    func recommendTools(recentTools: [String], topK: Int = 5) -> [ToolRecommendation] {
        if toolFeatures.isEmpty {
            buildToolFeatures()
        }

        guard !recentTools.isEmpty else {
            logger.info("[ContentRecommender] No recent tools for recommendation")
            return []
        }

        var toolScores: [String: (totalScore: Float, count: Int, basedOnTools: [(tool: String, similarity: Float)])] = [:]

        // Calculate similarity-based scores
        for recentTool in recentTools {
            let similarTools = findSimilarTools(recentTool)

            for similar in similarTools {
                // Skip if tool is in recent tools
                if recentTools.contains(similar.tool) { continue }

                if toolScores[similar.tool] == nil {
                    toolScores[similar.tool] = (totalScore: 0, count: 0, basedOnTools: [])
                }

                toolScores[similar.tool]?.totalScore += similar.similarity
                toolScores[similar.tool]?.count += 1
                toolScores[similar.tool]?.basedOnTools.append((tool: recentTool, similarity: similar.similarity))
            }
        }

        // Build recommendations
        var recommendations: [ToolRecommendation] = []

        for (tool, data) in toolScores {
            let avgScore = data.totalScore / Float(data.count)
            let confidence = min(Float(data.count) / Float(recentTools.count), 1.0)

            recommendations.append(ToolRecommendation(
                tool: tool,
                score: avgScore,
                confidence: confidence,
                reason: generateReason(basedOn: data.basedOnTools),
                algorithm: .contentBased
            ))
        }

        recommendations.sort { $0.score > $1.score }

        // Update stats
        stats.totalRecommendations += 1
        if let first = recommendations.first {
            let n = Float(stats.totalRecommendations)
            stats.avgSimilarity = (stats.avgSimilarity * (n - 1) + first.score) / n
        }

        return Array(recommendations.prefix(topK))
    }

    /// Tool chain recommendation (predict next tool)
    /// - Parameters:
    ///   - currentTool: Current tool being used
    ///   - topK: Number of recommendations
    /// - Returns: List of recommendations
    func recommendNextTools(currentTool: String, topK: Int = 3) -> [ToolRecommendation] {
        guard let nextTools = toolChains[currentTool], !nextTools.isEmpty else {
            return []
        }

        let recommendations = Array(nextTools.prefix(topK)).map { entry in
            ToolRecommendation(
                tool: entry.nextTool,
                score: entry.probability,
                confidence: entry.probability,
                reason: String(format: "%.0f%% of users use after %@", entry.probability * 100, currentTool),
                algorithm: .toolChain
            )
        }

        stats.toolChainRecommendations += 1

        return recommendations
    }

    /// Get statistics
    func getStats() -> [String: Any] {
        return [
            "totalRecommendations": stats.totalRecommendations,
            "toolChainRecommendations": stats.toolChainRecommendations,
            "toolCount": toolFeatures.count,
            "toolChainCount": toolChains.count,
            "avgSimilarity": String(format: "%.1f%%", stats.avgSimilarity * 100)
        ]
    }

    // MARK: - Private Methods

    /// Create feature vector for a tool
    private func createFeatureVector(category: ToolCategory, tags: [String]) -> [Float] {
        var vector: [Float] = []

        // 1. Category encoding (one-hot)
        for cat in ToolCategory.allCases {
            vector.append(cat == category ? 1.0 : 0.0)
        }

        // 2. Tag features (presence in standard tags)
        let tagSet = Set(tags)
        for standardTag in standardTags {
            vector.append(tagSet.contains(standardTag) ? 1.0 : 0.0)
        }

        return vector
    }

    /// Calculate cosine similarity
    private func cosineSimilarity(_ vec1: [Float], _ vec2: [Float]) -> Float {
        guard vec1.count == vec2.count, !vec1.isEmpty else { return 0 }

        var dotProduct: Float = 0
        var norm1: Float = 0
        var norm2: Float = 0

        for i in 0..<vec1.count {
            dotProduct += vec1[i] * vec2[i]
            norm1 += vec1[i] * vec1[i]
            norm2 += vec2[i] * vec2[i]
        }

        guard norm1 > 0, norm2 > 0 else { return 0 }

        return dotProduct / (sqrt(norm1) * sqrt(norm2))
    }

    /// Generate recommendation reason
    private func generateReason(basedOn tools: [(tool: String, similarity: Float)]) -> String {
        guard let topTool = tools.first else {
            return "Based on content similarity"
        }

        let similarity = Int(topTool.similarity * 100)
        return "Similar to \(topTool.tool) (\(similarity)%)"
    }
}
